site/content/blog/2021-11-17-database-sql.md

```diff
@@ -0,0 +1,219 @@
+
+```go
+// Compile-time check to ensure Connector implements the interface.
+_ driver.Connector = &Connector{}
+
+// NewConnector returns a DB connector that can be used to return a sql.DB object.
+// It takes a plugin.Driver implementation.
+func NewConnector(api plugin.Driver) *Connector {
+    return &Connector{api: api}
+}
+```
+
+The way it would work is to assign ids to each object (connection, statement, transaction, rows) in the `database/sql` universe in the client side, and track those objects in a map on the server.
+
+Here is a brief snippet of the `plugin.Driver` interface:
+
+```go
+type Driver interface {
+    Conn(isMaster bool) (string, error)
+    ConnPing(connID string) error
+    ConnClose(connID string) error
+    ConnQuery(connID, q string, args []driver.NamedValue) (string, error)
+    ConnExec(connID, q string, args []driver.NamedValue) (ResultContainer, error)
+    ...
+}
+```
+
+So a `*Conn.Ping` method on the client would call the `ConnPing` API and pass the connection ID. On the server side, it looked something like this:
+
+```go
+type DriverImpl struct {
+    s       *Server
+    connMut sync.RWMutex
+    connMap map[string]*sql.Conn
+    txMut   sync.Mutex
+    txMap   map[string]driver.Tx
+    stMut   sync.RWMutex
+    stMap   map[string]driver.Stmt
+    rowsMut sync.RWMutex
+    rowsMap map[string]driver.Rows
+}
+
+func (d *DriverImpl) ConnPing(connID string) error {
+    d.connMut.RLock()
+    conn, ok := d.connMap[connID]
+    d.connMut.RUnlock()
+    if !ok {
+        return driver.ErrBadConn
+    }
+
+    // Now use the conn to make the ping and return appropriately.
+}
+```
+
+### Roadblocks
+
+**1. Raw access to `driver.Conn`**
+
+The first problem we hit was to get access to the raw `driver.Conn` instance. Since the client was working at the `database/sql/driver` level, once we would get the query params on the server side, we cannot go via the `database/sql` layer again! We needed access to the underlying `driver.Conn` object to directly call the driver APIs.
+
+Fortunately, the [*Conn.Raw](https://pkg.go.dev/database/sql#Conn.Raw) method was there to save the day!
+
+And so this is how the rest of `ConnPing` actually looks:
+
+```go
+return conn.Raw(func(innerConn interface{}) error {
+    return innerConn.(driver.Pinger).Ping(context.Background())
+})
+```
+
+**2. Cannot pass `context.Context`**
+
+The second problem was that since `context.Context` was an interface with an unexported underlying struct, we could not pass that, because on the server side, the Go unmarshaler did not know the underlying implementation of that. So as a result we could not pass information like timeout values and such. While this was solvable, it did not affect us that much because the timeout values were the same for all queries and we just set them on the server side.
```

Comment:
nit: Generally the term "serialization" is used for the intent of marshaling. But 0/5, maybe it's too java-ish 🤷‍♂️

Reply:
The API uses the term marshaling: https://pkg.go.dev/encoding#BinaryUnmarshaler, so I stuck with it :P